codegenerator/cli/templates/static/codegen/src/db/Table.res

```diff
@@ -186,3 +198,53 @@ let getCompositeIndices = (table): array<array<string>> => {
   ->getUnfilteredCompositeIndicesUnsafe
   ->Array.keep(ind => ind->Array.length > 1)
 }
+
+module PostgresInterop = {
+  type pgFn<'payload, 'return> = (Postgres.sql, 'payload) => promise<'return>
+  type batchSetFn<'a> = (Postgres.sql, array<'a>) => promise<unit>
+  external eval: string => 'a = "eval"
+
+  let makeBatchSetFnString = (table: table) => {
+    let fieldNamesInQuotes =
+      table->getNonDefaultFieldNames->Array.map(fieldName => `"${fieldName}"`)
+    `(sql, rows) => {
+      return sql\`
+        INSERT INTO "public"."${table.tableName}"
+        \${sql(rows, ${fieldNamesInQuotes->Js.Array2.joinWith(", ")})}
+        ON CONFLICT(${table->getPrimaryKeyFieldNames->Js.Array2.joinWith(", ")}) DO UPDATE
+        SET
+        ${fieldNamesInQuotes
+      ->Array.map(fieldNameInQuotes => `${fieldNameInQuotes} = EXCLUDED.${fieldNameInQuotes}`)
+      ->Js.Array2.joinWith(", ")};\`
+    }`
+  }
+
+  let chunkBatchQuery = async (
+    sql,
+    entityDataArray: array<'entity>,
+    queryToExecute: pgFn<array<'entity>, 'return>,
+    ~maxItemsPerQuery=500,
+  ) => {
+    let responses = []
+    let i = ref(0)
+    let shouldContinue = () => i.contents < entityDataArray->Array.length
+    // Split entityDataArray into chunks of maxItemsPerQuery
+    while shouldContinue() {
+      let chunk =
+        entityDataArray->Js.Array2.slice(~start=i.contents, ~end_=i.contents + maxItemsPerQuery)
+      let response = await queryToExecute(sql, chunk)
+      responses->Js.Array2.push(response)->ignore
+      i := i.contents + maxItemsPerQuery
+    }
+    responses
+  }
+
+  let makeBatchSetFn = (~table, ~rowsSchema: S.t<array<'a>>): batchSetFn<'a> => {
+    let batchSetFn: pgFn<array<Js.Json.t>, unit> = table->makeBatchSetFnString->eval
+    async (sql, rows) => {
+      let rowsJson =
+        rows->S.serializeOrRaiseWith(rowsSchema)->(Utils.magic: Js.Json.t => array<Js.Json.t>)
+      let _res = await chunkBatchQuery(sql, rowsJson, batchSetFn)
+    }
+  }
+}
```

Comment:
> Is there any movement in the Rescript community of binding to these kinds of template literals like `sql`` directly?

There already is an implementation but it only works for if you can bind directly to the the template literal function. In this case we instantiate `sql` first and only then it can be used as an object literal. 

There's a few more problems that `eval` solves here rather than just templating a query though. 

If you like in the DbFunctionsImplementation.js, I have already found a way to generalize queries using just the library and this is currently working for all codegen entities. These problem with it is, it's super fragile code, really difficult to read or reason about and all they `dynamicness` is offloaded to the runtime of the function so there is some runtime overhead. With eval, we essentially JIT compile a static function with no extra runtime overhead. The resulting query to postgres is much simpler and the js function becomes more testable.

Reply:
> Hmm, I'm thinking it might be better still to generate these functions at codegen time rather than with eval in this case. 

I would disagree that it is better to codegen these things. This is exactly what we used to do with all the entity insert functions etc at codegen time. One of the problems is that there is no relationship between hardcoded codegen values and the frameworks being built. So for eg. say I make an edit to  the `Table.t` type and how it gets used. Those changes don't trickle down to the db interop without changing hardcoded fields. Say for instance I change the `getFields` function on the table to filter out defaulted fields or something. Now I need to rewrite that logic in the codegenerator and there's no relationship to the framework.

> Is the primary motivation that you want everything to just be a library and you just set some parameters once on this JavaScript library and everything gets generated in code rather than codegen step?

I would say yes, the main motivation is that we are trying to move away from leaning on codegen and make the whole indexer more lib like. But the codegen still has the ultimate responsibility of generating types and apis while everything in the lib should tend towards static reusable code.

Reply:
Cool sounds good. I was just poking a little bit to understand better. :+1: